File: java/src/main/java/ps_traning/programmers/level_3/이중우선순위큐.java
package ps_traning.programmers.level_3;

import java.util.Collections;
import java.util.PriorityQueue;

public class 이중우선순위큐 {
    public int[] solution(String[] operations) {
        int[] answer = {0, 0};
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        PriorityQueue<Integer> rpq = new PriorityQueue<>(Collections.reverseOrder());

        for (String op : operations) {
            String[] temp = op.split(" ");
            String command = temp[0];
            int num = Integer.valueOf(temp[1]);

            if (command.equals("I")) {
                pq.offer(num);
                rpq.offer(num);
            } else if (command.equals("D")) {
                if (num == -1 && !pq.isEmpty()) {
                    rpq.remove(pq.poll());
                } else if (num == 1 && !rpq.isEmpty()) {
                    pq.remove(rpq.poll());
                }
            }
        }

        if (!pq.isEmpty()) {
            answer[0] = rpq.peek();
            answer[1] = pq.peek();
        }

        return answer;
    }
}

File: python/with_java/programmers/level 3/이중우선순위큐.py
import heapq

def solution(operations):
    heap = []
    status = -1
    for op in operations:
        command, num = op.split()
        num = int(num)
        if command == "I":
            if status == 1:
                num = -num
            heapq.heappush(heap, num)
        if command == "D":
            if num != status:
                heap = list(map(lambda x:-x, heap))
                heapq.heapify(heap)
                status = num
            if heap:
                heapq.heappop(heap)
    answer = []
    if heap:
        if status == 1:
            answer.append(-heap[0])
            heap = list(map(lambda x:-x, heap))
            heapq.heapify(heap)
            answer.append(heap[0])
        else:
            answer.append(heap[0])
            heap = list(map(lambda x:-x, heap))
            heapq.heapify(heap)
            answer.append(-heap[0])
        answer.sort(reverse=True)
    else:
        answer = [0, 0]
            
    return answer